File: icamera/src/main/java/me/shouheng/icamera/util/ImageHelper.java
package me.shouheng.icamera.util;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.Rect;
import android.graphics.YuvImage;
import android.media.Image;
import android.os.Build;
import android.support.annotation.RequiresApi;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ReadOnlyBufferException;

public final class ImageHelper {

    private static final String TAG = "ImageHelper";

    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    public static byte[] convertYUV_420_888toNV21(Image image) {
        int width = image.getWidth();
        int height = image.getHeight();
        int ySize = width*height;
        int uvSize = width*height/4;

        byte[] nv21 = new byte[ySize + uvSize*2];

        ByteBuffer yBuffer = image.getPlanes()[0].getBuffer(); // Y
        ByteBuffer uBuffer = image.getPlanes()[1].getBuffer(); // U
        ByteBuffer vBuffer = image.getPlanes()[2].getBuffer(); // V

        int rowStride = image.getPlanes()[0].getRowStride();

        int pos = 0;

        if (rowStride == width) { // likely
            yBuffer.get(nv21, 0, ySize);
            pos += ySize;
        } else {
            int yBufferPos = -rowStride; // not an actual position
            for (; pos<ySize; pos+=width) {
                yBufferPos += rowStride;
                yBuffer.position(yBufferPos);
                yBuffer.get(nv21, pos, width);
            }
        }

        rowStride = image.getPlanes()[2].getRowStride();
        int pixelStride = image.getPlanes()[2].getPixelStride();

        if (pixelStride == 2 && rowStride == width && uBuffer.get(0) == vBuffer.get(1)) {
            // maybe V an U planes overlap as per NV21, which means vBuffer[1] is alias of uBuffer[0]
            byte savePixel = vBuffer.get(1);
            try {
                vBuffer.put(1, (byte) ~savePixel);
                if (uBuffer.get(0) == (byte) ~savePixel) {
                    vBuffer.put(1, savePixel);
                    vBuffer.get(nv21, ySize, uvSize);

                    return nv21; // shortcut
                }
            } catch (ReadOnlyBufferException ex) {
                XLog.e(TAG, "ReadOnlyBufferException :" + ex);
            }
            vBuffer.put(1, savePixel);
        }

        for (int row=0, row_len=height/2; row<row_len; row++) {
            for (int col=0, col_len=width/2; col<col_len; col++) {
                int vuPos = col*pixelStride + row*rowStride;
                nv21[pos++] = vBuffer.get(vuPos);
                nv21[pos++] = uBuffer.get(vuPos);
            }
        }
        return nv21;
    }

    public static int convertYUV420_NV21toARGB8888(byte[] data, int width, int height) {
        int size = width * height;
        int[] bri = new int[4];
        int u, v, y1, y2, y3, y4;

        // i along Y and the final pixels
        // k along pixels U and V
        int i = 0;
        y1 = data[i] & 0xff;
        y2 = data[i + 1] & 0xff;
        y3 = data[width + i] & 0xff;
        y4 = data[width + i + 1] & 0xff;

        bri[0] = y1;
        bri[1] = y2;
        bri[2] = y3;
        bri[3] = y4;

        int max = 0;
        for (int j = 0; i < bri.length; i++) {
            if (bri[j] > max) {
                max = bri[j];
            }
        }
        return max;
    }

    public static Bitmap convertNV21ToBitmap(byte[] nv21, int width, int height) {
        Bitmap bitmap = null;
        try {
            YuvImage image = new YuvImage(nv21, ImageFormat.NV21, width, height, null);
            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            image.compressToJpeg(new Rect(0, 0, width, height), 80, stream);
            bitmap = BitmapFactory.decodeByteArray(stream.toByteArray(), 0, stream.size());
            stream.close();
        } catch (IOException e) {
            e.printStackTrace();
            XLog.e(TAG, "nv21ToBitmap : " + e);
        }
        return bitmap;
    }

    private ImageHelper() {
        throw new UnsupportedOperationException("u can't initialize me!");
    }
}

File: icamera/src/main/java/me/shouheng/icamera/util/ImageHelper.kt
package me.shouheng.icamera.util

import android.graphics.*
import android.media.Image
import android.os.Build
import androidx.annotation.RequiresApi
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.ReadOnlyBufferException
import kotlin.experimental.and
import kotlin.experimental.inv

object ImageHelper {

    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    fun convertYUV_420_888toNV21(image: Image): ByteArray {
        val width = image.width
        val height = image.height
        val ySize = width * height
        val uvSize = width * height / 4
        val nv21 = ByteArray(ySize + uvSize * 2)
        val yBuffer = image.planes[0].buffer // Y
        val uBuffer = image.planes[1].buffer // U
        val vBuffer = image.planes[2].buffer // V
        var rowStride = image.planes[0].rowStride
        var pos = 0
        if (rowStride == width) { // likely
            yBuffer[nv21, 0, ySize]
            pos += ySize
        } else {
            var yBufferPos = -rowStride // not an actual position
            while (pos < ySize) {
                yBufferPos += rowStride
                yBuffer.position(yBufferPos)
                yBuffer[nv21, pos, width]
                pos += width
            }
        }
        rowStride = image.planes[2].rowStride
        val pixelStride = image.planes[2].pixelStride
        if (pixelStride == 2 && rowStride == width && uBuffer[0] == vBuffer[1]) {
            // maybe V an U planes overlap as per NV21, which means vBuffer[1] is alias of uBuffer[0]
            val savePixel = vBuffer[1]
            try {
                vBuffer.put(1, savePixel.inv())
                if (uBuffer[0] == savePixel.inv()) {
                    vBuffer.put(1, savePixel)
                    vBuffer[nv21, ySize, uvSize]
                    return nv21 // shortcut
                }
            } catch (ex: ReadOnlyBufferException) {
                XLog.e("ImageHelper", "ReadOnlyBufferException :$ex")
            }
            vBuffer.put(1, savePixel)
        }
        var row = 0
        val row_len = height / 2
        while (row < row_len) {
            var col = 0
            val col_len = width / 2
            while (col < col_len) {
                val vuPos = col * pixelStride + row * rowStride
                nv21[pos++] = vBuffer[vuPos]
                nv21[pos++] = uBuffer[vuPos]
                col++
            }
            row++
        }
        return nv21
    }

    fun convertYUV420_NV21toARGB8888(data: ByteArray, width: Int, height: Int): Int {
        val size = width * height
        val bri = IntArray(4)
        var u: Int
        var v: Int
        val y1: Int
        val y2: Int
        val y3: Int
        val y4: Int

        // i along Y and the final pixels
        // k along pixels U and V
        var i = 0
        y1 = (data[i] and 0xff.toByte()).toInt()
        y2 = (data[i + 1] and 0xff.toByte()).toInt()
        y3 = (data[width + i] and 0xff.toByte()).toInt()
        y4 = (data[width + i + 1] and 0xff.toByte()).toInt()
        bri[0] = y1
        bri[1] = y2
        bri[2] = y3
        bri[3] = y4
        var max = 0
        val j = 0
        while (i < bri.size) {
            if (bri[j] > max) {
                max = bri[j]
            }
            i++
        }
        return max
    }

    fun convertNV21ToBitmap(nv21: ByteArray?, width: Int, height: Int): Bitmap? {
        var bitmap: Bitmap? = null
        try {
            val image = YuvImage(nv21, ImageFormat.NV21, width, height, null)
            val stream = ByteArrayOutputStream()
            image.compressToJpeg(Rect(0, 0, width, height), 80, stream)
            bitmap = BitmapFactory.decodeByteArray(stream.toByteArray(), 0, stream.size())
            stream.close()
        } catch (e: IOException) {
            XLog.e("ImageHelper", "nv21ToBitmap : $e")
        }
        return bitmap
    }
}